src/sas/webfit/analyze/fitting/tests.py

```diff
@@ -0,0 +1,148 @@
+from django.shortcuts import get_object_or_404
+from rest_framework import status
+from rest_framework.test import APIRequestFactory, APIClient, force_authenticate, APITestCase
+
+from sasmodels.data import empty_data1D
+from sasdata.dataloader.loader import Loader
+import numpy as np
+
+from data.models import Data
+from .models import (
+    Fit,
+    FitParameter
+)
+from .views import (
+    start,
+    start_fit,
+)
+
+factory = APIRequestFactory()
+
+# Create your tests here.
+class TestLists(APITestCase):
+    """def setUp(self):
+        self.user = User.objects.create_user(username="testUser", password="secret", id = 2)
+        self.client = APIClient()
+        self.client.force_authenticate(user=self.user)"""
+
+    #working
+    def get_model_list(self):
+        request = self.client.post('/v1/analyze/fit/models/')
+        self.assertEqual(request.data, {"all models":['adsorbed_layer', 'barbell', 'bcc_paracrystal', 'be_polyelectrolyte', 'binary_hard_sphere', 'broad_peak', 'capped_cylinder', 'core_multi_shell', 'core_shell_bicelle', 'core_shell_bicelle_elliptical', 'core_shell_bicelle_elliptical_belt_rough', 'core_shell_cylinder', 'core_shell_ellipsoid', 'core_shell_parallelepiped', 'core_shell_sphere', 'correlation_length', 'cylinder', 'dab', 'ellipsoid', 'elliptical_cylinder', 'fcc_paracrystal', 'flexible_cylinder', 'flexible_cylinder_elliptical', 'fractal', 'fractal_core_shell', 'fuzzy_sphere', 'gauss_lorentz_gel', 'gaussian_peak', 'gel_fit', 'guinier', 'guinier_porod', 'hardsphere', 'hayter_msa', 'hollow_cylinder', 'hollow_rectangular_prism', 'hollow_rectangular_prism_thin_walls', 'lamellar', 'lamellar_hg', 'lamellar_hg_stack_caille', 'lamellar_stack_caille', 'lamellar_stack_paracrystal', 'line', 'linear_pearls', 'lorentz', 'mass_fractal', 'mass_surface_fractal', 'mono_gauss_coil', 'multilayer_vesicle', 'onion', 'parallelepiped', 'peak_lorentz', 'pearl_necklace', 'poly_gauss_coil', 'polymer_excl_volume', 'polymer_micelle', 'porod', 'power_law', 'pringle', 'raspberry', 'rectangular_prism', 'rpa', 'sc_paracrystal', 'sphere', 'spherical_sld', 'spinodal', 'squarewell', 'stacked_disks', 'star_polymer', 'stickyhardsphere', 'superball', 'surface_fractal', 'teubner_strey', 'triaxial_ellipsoid', 'two_lorentzian', 'two_power_law', 'unified_power_Rg', 'vesicle']})
+    
+    #working
+    def get_model_list_category(self):
+        data = {
+            "category":"Cylinder"
+        }
+        request = self.client.post('/v1/analyze/fit/models/', data, format='json')
+        self.assertEqual(request.data, {"Cylinder models":[['barbell', True], ['capped_cylinder', True], ['core_shell_bicelle', True], ['core_shell_bicelle_elliptical', True], ['core_shell_bicelle_elliptical_belt_rough', True], ['core_shell_cylinder', True], ['cylinder', True], ['elliptical_cylinder', True], ['flexible_cylinder', True], ['flexible_cylinder_elliptical', True], ['hollow_cylinder', True], ['pearl_necklace', True], ['pringle', True], ['stacked_disks', True]]})
```

Comment:
For now, Django REST framework only accepts data from user if the request is POST or PUT. We were wondering if we should have moved the data to URL, but ultimately decided that it was an argument about semantics ("is it really a GET if the user is giving data, even through URL?"). 

Plus there are benefits to using POST as in later implementations, as the User might specify they want both category and kind. However, currently, I could easily change it GET and only use URL. What do you think is better?

The reason why fit/models/ data = "category":"cylinder" uses CategoryInstaller, which (I believe) checks if each model is True. Should I find another way to get categories, or should I reformat the list to only get the string/[0]?

The format = 'json' refers to the format of the data for the "client" and is only necessary for Django tests. This is because the Django tests "client's" data is originally a python dictionary, and is turned into a string, rather than a json (which the final API takes).

Reply:
Webfit should be a thin wrapper around the services provided by sascalc/sasdata/sasmodels. Do as little as possible to the data you return.

Chances are qtgui and webfit can use the same interface. For categories I see that you have reimplemented the same code for loading the category file that appears in multiple places in qtgui. All of these should be using the same function so we can maintain it in one place. I suggest writing a function to load the categories into a {category: [model, ...]} data structure so that your model view function becomes trivial. Later we can move it to the plugin manager code in sascalc and restructure qt gui to use it.

One complication is that gui allows users to modify categories. This implies that we should have a category list for each user of webfit and an inteface for maintaining them. Ignore this for now.

Another complication is that the GUI is filtering out the rpa model whereever it appears. Maintain a blacklist along with the category loader code so that the GUI and webfit will present the same set of models.

Reply:
Re: GET vs POST, it is generally easier to use GET. You can simply type the URL into a browser and see the result. For example, 

* the first page of sasview issues: https://api.github.com/repos/SasView/sasview/issues
* details on a particular issue: https://api.github.com/repos/SasView/sasview/issues/2583

Neither of these require any data to be posted because the url itself contains the data. They also don't require authentication because this is public facing information. This may not be the case for webfit.